### PR TITLE
Simplify shape-related operations

### PR DESCRIPTION
**Pull request by tungld**
This patch is to simplify shape-related operations that are often generated by onnx converters and complicate our shape inference for unknown dimensions.

Usually shape-related calculation starts with an ONNXShapeOp to get the shape of a tensor, e.g.
```mlir
%0 = "onnx.Shape"(%arg0) : (tensor<?x256x1xf32>) -> tensor<3xi64>
%1 = "onnx.Reshape"(%arg1, %0) : (tensor<?x256xf32>,  tensor<3xi64>) -> tensor<?x?x?xf32>
```
And the problem is that ONNXShapeOp returns a tensor whose values are unknown at compile time, so the value `256` in <?x256> cannot be propagated further after this ONNXShapeOp. Hence, the output shape of Reshape is totally unknown.

The essential idea in this patch to solve the problem is representing ONNXShapeOp by ONNXDimOp (newly introduced in this path) and ONNXConcatOp. By that way, each invidual dimension of a tensor is exposed and easily propagated through other ops. For example, the above "onnx.Shape" will rewritten into:
```mlir
%0 = "onnx.Dim"(%arg0) {axis = 0 : si64} : (tensor<?x256x1xi64>) -> tensor<1xi64>
%1 = "onnx.Constant"() {value = dense<256> : tensor<1xi64>} : () -> tensor<1xi64>
%2 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
%3 = "onnx.Concat"(%0, %1, %2) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<3xi64>
%4 = "onnx.Reshape"(%arg0, %3) : (tensor<?x256xf32>,  tensor<3xi64>) -> tensor<?x?x?xf32>
``` 

Now, it's straighforward to update the output shape of Reshape from `<?x?x?xf32>` to `<?x256x1xf32>` by looking at the inputs of Concat.

Applying this patch to an example in bertsquad-12 model (See #1579), the input program:
```mlir
module attributes {llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-unknown-linux-gnu", "onnx-mlir.accels" = ["NNPA-0x10000"]} {
  func.func @main_graph(%arg0: tensor<?xi64>, %arg1: tensor<?x256xi64>, %arg2: tensor<?x256xi64>, %arg3: tensor<?x256xi64>) -> (tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>) attributes {input_names = ["unique_ids_raw_output___9:0", "segment_ids:0", "input_mask:0", "input_ids:0"], output_names = ["unstack:1", "unstack:0", "unique_ids:0"]} {
    %0 = "onnx.Constant"() {value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
    %1 = "onnx.Constant"() {value = dense<256> : tensor<1xi32>} : () -> tensor<1xi32>
    %2 = "onnx.Shape"(%arg3) {onnx_node_name = "bert/encoder/Shape"} : (tensor<?x256xi64>) -> tensor<2xi64>
    %3 = "onnx.Cast"(%2) {onnx_node_name = "bert/encoder/Shape__12", to = f32} : (tensor<2xi64>) -> tensor<2xf32>
    %4 = "onnx.Constant"() {value = dense<0> : tensor<1xi64>} : () -> tensor<1xi64>
    %5 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
    %6 = "onnx.Constant"() {value = dense<0> : tensor<1xi64>} : () -> tensor<1xi64>
    %7 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
    %8 = "onnx.Slice"(%3, %4, %5, %6, %7) {onnx_node_name = "bert/encoder/strided_slice"} : (tensor<2xf32>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<1xf32>
    %9 = "onnx.SqueezeV11"(%8) {axes = [0], onnx_node_name = "bert/encoder/strided_slice__16"} : (tensor<1xf32>) -> tensor<f32>
    %10 = "onnx.Cast"(%9) {onnx_node_name = "bert/encoder/strided_slice__17", to = i32} : (tensor<f32>) -> tensor<i32>
    %11 = "onnx.UnsqueezeV11"(%10) {axes = [0], onnx_node_name = "bert/encoder/ones/packed_Unsqueeze__18"} : (tensor<i32>) -> tensor<1xi32>
    %12 = "onnx.Concat"(%11, %1, %0) {axis = 0 : si64, onnx_node_name = "bert/encoder/ones/packed_Concat__21"} : (tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<3xi32>
    %13 = "onnx.Cast"(%12) {onnx_node_name = "bert/encoder/ones__22", to = i64} : (tensor<3xi32>) -> tensor<3xi64>
    %14 = "onnx.ConstantOfShape"(%13) {onnx_node_name = "bert/encoder/ones", value = dense<1.000000e+00> : tensor<1xf32>} : (tensor<3xi64>) -> tensor<?x?x?xf32>
    %15 = "onnx.UnsqueezeV11"(%10) {axes = [0], onnx_node_name = "bert/encoder/Reshape/shape_Unsqueeze__23"} : (tensor<i32>) -> tensor<1xi32>
    %16 = "onnx.Constant"() {value = dense<256> : tensor<1xi32>} : () -> tensor<1xi32>
    %17 = "onnx.Constant"() {value = dense<1> : tensor<1xi32>} : () -> tensor<1xi32>
    %18 = "onnx.Concat"(%15, %17, %16) {axis = 0 : si64, onnx_node_name = "bert/encoder/Reshape/shape_Concat__26"} : (tensor<1xi32>, tensor<1xi32>, tensor<1xi32>) -> tensor<3xi32>
    %19 = "onnx.Cast"(%18) {onnx_node_name = "bert/encoder/Reshape__27", to = i64} : (tensor<3xi32>) -> tensor<3xi64>
    %20 = "onnx.Reshape"(%arg2, %19) {onnx_node_name = "bert/encoder/Reshape"} : (tensor<?x256xi64>, tensor<3xi64>) -> tensor<?x?x?xi64>
    %21 = "onnx.Cast"(%20) {onnx_node_name = "bert/encoder/Cast", to = f32} : (tensor<?x?x?xi64>) -> tensor<?x?x?xf32>
    %22 = "onnx.Mul"(%14, %21) {onnx_node_name = "bert/encoder/mul"} : (tensor<?x?x?xf32>, tensor<?x?x?xf32>) -> tensor<?x?x?xf32>
    %23 = "onnx.Constant"() {value = dense<[-1, 1, 256, 256]> : tensor<4xi64>} : () -> tensor<4xi64>
    %24 = "onnx.Reshape"(%22, %23) {onnx_node_name = "bert/encoder/layer_1/attention/self/ExpandDims"} : (tensor<?x?x?xf32>, tensor<4xi64>) -> tensor<?x1x256x256xf32>
    %25 = "onnx.Constant"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %26 = "onnx.Sub"(%25, %24) {onnx_node_name = "bert/encoder/layer_1/attention/self/sub"} : (tensor<f32>, tensor<?x1x256x256xf32>) -> tensor<?x1x256x256xf32>
    %27 = "onnx.Constant"() {value = dense<-1.000000e+04> : tensor<f32>} : () -> tensor<f32>
    %28 = "onnx.Mul"(%26, %27) {onnx_node_name = "bert/encoder/layer_1/attention/self/mul_1"} : (tensor<?x1x256x256xf32>, tensor<f32>) -> tensor<?x1x256x256xf32>
    %29 = "onnx.Constant"() {value = dense<[-1, 1, 256, 256]> : tensor<4xi64>} : () -> tensor<4xi64>
    %30 = "onnx.Reshape"(%22, %29) {onnx_node_name = "bert/encoder/layer_0/attention/self/ExpandDims"} : (tensor<?x?x?xf32>, tensor<4xi64>) -> tensor<?x1x256x256xf32>
    %31 = "onnx.Constant"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %32 = "onnx.Sub"(%31, %30) {onnx_node_name = "bert/encoder/layer_0/attention/self/sub"} : (tensor<f32>, tensor<?x1x256x256xf32>) -> tensor<?x1x256x256xf32>
    %33 = "onnx.Constant"() {value = dense<-1.000000e+04> : tensor<f32>} : () -> tensor<f32>
    %34 = "onnx.Mul"(%32, %33) {onnx_node_name = "bert/encoder/layer_0/attention/self/mul_1"} : (tensor<?x1x256x256xf32>, tensor<f32>) -> tensor<?x1x256x256xf32>
    return %34, %32, %28 : tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph} : () -> ()
}
```
will become
```mlir
module attributes {llvm.data_layout = "E-m:e-i1:8:16-i8:8:16-i64:64-f128:64-a:8:16-n32:64", llvm.target_triple = "s390x-ibm-linux", "onnx-mlir.accels" = ["NNPA-0x10000"]} {
  func.func @main_graph(%arg0: tensor<?xi64>, %arg1: tensor<?x256xi64>, %arg2: tensor<?x256xi64>, %arg3: tensor<?x256xi64>) -> (tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>) attributes {input_names = ["unique_ids_raw_output___9:0", "segment_ids:0", "input_mask:0", "input_ids:0"], output_names = ["unstack:1", "unstack:0", "unique_ids:0"]} {
    %0 = "onnx.Dim"(%arg3) {axis = 0 : si64} : (tensor<?x256xi64>) -> tensor<1xi64>
    %1 = "onnx.Constant"() {value = dense<256> : tensor<1xi64>} : () -> tensor<1xi64>
    %2 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
    %3 = "onnx.Concat"(%0, %1, %2) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<3xi64>
    %4 = "onnx.ConstantOfShape"(%3) {onnx_node_name = "bert/encoder/ones", value = dense<1.000000e+00> : tensor<1xf32>} : (tensor<3xi64>) -> tensor<?x256x1xf32>
    %5 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
    %6 = "onnx.Constant"() {value = dense<256> : tensor<1xi64>} : () -> tensor<1xi64>
    %7 = "onnx.Concat"(%0, %5, %6) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<3xi64>
    %8 = "onnx.Reshape"(%arg2, %7) {onnx_node_name = "bert/encoder/Reshape"} : (tensor<?x256xi64>, tensor<3xi64>) -> tensor<?x1x256xi64>
    %9 = "onnx.Cast"(%8) {onnx_node_name = "bert/encoder/Cast", to = f32} : (tensor<?x1x256xi64>) -> tensor<?x1x256xf32>
    %10 = "onnx.Mul"(%4, %9) {onnx_node_name = "bert/encoder/mul"} : (tensor<?x256x1xf32>, tensor<?x1x256xf32>) -> tensor<?x256x256xf32>
    %11 = "onnx.Constant"() {value = dense<[-1, 1, 256, 256]> : tensor<4xi64>} : () -> tensor<4xi64>
    %12 = "onnx.Reshape"(%10, %11) {onnx_node_name = "bert/encoder/layer_1/attention/self/ExpandDims"} : (tensor<?x256x256xf32>, tensor<4xi64>) -> tensor<?x1x256x256xf32>
    %13 = "onnx.Constant"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %14 = "onnx.Sub"(%13, %12) {onnx_node_name = "bert/encoder/layer_1/attention/self/sub"} : (tensor<f32>, tensor<?x1x256x256xf32>) -> tensor<?x1x256x256xf32>
    %15 = "onnx.Constant"() {value = dense<-1.000000e+04> : tensor<f32>} : () -> tensor<f32>
    %16 = "onnx.Mul"(%14, %15) {onnx_node_name = "bert/encoder/layer_1/attention/self/mul_1"} : (tensor<?x1x256x256xf32>, tensor<f32>) -> tensor<?x1x256x256xf32>
    %17 = "onnx.Constant"() {value = dense<[-1, 1, 256, 256]> : tensor<4xi64>} : () -> tensor<4xi64>
    %18 = "onnx.Reshape"(%10, %17) {onnx_node_name = "bert/encoder/layer_0/attention/self/ExpandDims"} : (tensor<?x256x256xf32>, tensor<4xi64>) -> tensor<?x1x256x256xf32>
    %19 = "onnx.Constant"() {value = dense<1.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %20 = "onnx.Sub"(%19, %18) {onnx_node_name = "bert/encoder/layer_0/attention/self/sub"} : (tensor<f32>, tensor<?x1x256x256xf32>) -> tensor<?x1x256x256xf32>
    %21 = "onnx.Constant"() {value = dense<-1.000000e+04> : tensor<f32>} : () -> tensor<f32>
    %22 = "onnx.Mul"(%20, %21) {onnx_node_name = "bert/encoder/layer_0/attention/self/mul_1"} : (tensor<?x1x256x256xf32>, tensor<f32>) -> tensor<?x1x256x256xf32>
    return %22, %20, %16 : tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>, tensor<?x1x256x256xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph} : () -> ()
}
```
where all ops after Reshape gets more concrete shape information instead of `<?x?x?xf32>`.

Signed-off-by: Tung D. Le <tung@jp.ibm.com>

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
@tungld I like your reuse of the concat, essentially you are concatenating mixes of constants and dims directly, which works well. It is more direct than the original proposal in #1579, so a good improvement as we reuse more of the existing infrastructure.

For symbolic analysis, I think that we can use the SSA id of the dim (passed to the concat) and the constants to provide some info. Eventually, if there were computations on the dim, the result of this computation could be similarly passed to the concat and be treated as a new "unique" representation of a computed dim, so I think this would work well too. 

Bert had "annoying" conversions of shapes from `int` to `float` and back. Does your approach handle these smoothly? Would be interested to know how you dealt with that problem.

The only thing that we might be missing is that sometimes two dynamic dimensions are identical. This is obviously for another PR,  we could have a revised shape inference where negative number smaller than `-1` could represent an unique identifier setting the equality between dynamic dims. If param `X` is `(-2, 256, 256)` and `Y` is `(-1, -2, 3, 3)` indicating that the 1st dim of X is the same as the second dim of Y, we would have to build a dictionary of X/dim1 and Y/dim2 connecting these with a common Value, initially undefined. Then when asking for the dim of either, the first call would set the dict value to the created dim; further calls to dims of X/dim1 or Y/dim2 would then be always connected to that value. All further symbolic computations would then rely on the normal SSA as we artificially mapped these two seemingly unconnected dims to the same SSA value. This processing would be done when introducing the `ONNX.dim` op, which could be hidden in our ONNX dialect builder.

@chentong319 feel free to comment as well.

**Comment by tungld** [comment]
@AlexandreEichenberger thanks for your comments!

> Bert had "annoying" conversions of shapes from int to float and back. Does your approach handle these smoothly? Would be interested to know how you dealt with that problem.

Yes, it works well in this situation. Cast ops are moved towards the constant and dims as much as possible and then they will cancel each other or become new constants by constant propagation.

File: src/Conversion/ONNXToKrnl/Tensor/ConstantOfShape.cpp
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Tensor/ConstantOfShape.cpp:62]: This is to fix ConstantOfShape to load dynamic dims only.

File: src/Transform/ONNX/ConvOpt.cpp
Comment (tungld) [inline src/Transform/ONNX/ConvOpt.cpp:212]: This is to fix a compilation warning about pass constructor.

File: src/Transform/ONNX/Decompose.cpp
Comment (tungld) [inline src/Transform/ONNX/Decompose.cpp:208]: This is to fix a compilation warning about pass constructor.

**Comment by tungld** [comment]
@jenkins-droid test this please

**Comment by tungld** [comment]
@AlexandreEichenberger @chentong319 this patch is ready for review now. 

I will work up another PR for symbolic shape analytics later.

File: src/Dialect/ONNX/ONNX.td
Comment (AlexandreEichenberger) [inline src/Dialect/ONNX/ONNX.td:220]: I understand the convenience of using the ONNX dialect both for the standard ops as well as ops that are useful for us to perform some optimizations.

Would it make sense to label ops that we add for convenience (and I understand this is not the first one) as "Helper Ops" in the description? What do you think. As we add more of them, it might be nice to have a clear marker of the ones we added.

File: src/Transform/ONNX/SimplifyShapeRelatedOps.cpp
Comment (AlexandreEichenberger) [inline src/Transform/ONNX/SimplifyShapeRelatedOps.cpp:183]: PR #1684 is ready to land, and it adds the opset 15 support for start and end. Do you mind waiting for that PR to land and add support for this here? It would be much easier for you to do, as opposed for @philass to add it if yours land first. Thanks

**Review by AlexandreEichenberger — COMMENTED** [review]
Very good. I am asking you to let another PR that add support for slice with start/end (opset15) to go first so that you may quickly add that support to the optimization here, if it actually need special handling.
Otherwise, its all good.

File: src/Dialect/ONNX/ONNX.td
Comment (tungld) [reply src/Dialect/ONNX/ONNX.td:220]: Totally agree. Will add some markers for these ops.

File: src/Transform/ONNX/SimplifyShapeRelatedOps.cpp
Comment (tungld) [reply src/Transform/ONNX/SimplifyShapeRelatedOps.cpp:183]: Yes, it's absolutely fine. I'll wait for #1684.

Comment (philass) [reply src/Transform/ONNX/SimplifyShapeRelatedOps.cpp:183]: Thanks @tungld! I didn't realize you guys were waiting for me. Much appreciated

I'll fix the CI failures in my PR right away, and merge it into head once thats done.

Comment (AlexandreEichenberger) [reply src/Transform/ONNX/SimplifyShapeRelatedOps.cpp:183]: Thanks

File: src/Compiler/CompilerPasses.cpp
Comment (tungld) [reply src/Compiler/CompilerPasses.cpp:82]: It's currently self-contained which is easier to debug. It's possible to move it into the above loop when needed.

File: src/Dialect/ONNX/ONNX.td
Comment (tungld) [reply src/Dialect/ONNX/ONNX.td:225]: I added this sentent at the beginning of the description: `This operation is defined by onnx-mlir and not in the ONNX specification.`

File: src/Transform/ONNX/SimplifyShapeRelatedOps.cpp
Comment (tungld) [reply src/Transform/ONNX/SimplifyShapeRelatedOps.cpp:183]: @AlexandreEichenberger since #1684 was landed, I updated this patch to support ShapeOp opset 15.

@philass I found that the lowering of ShapeOp opset 15 failed if start is negative. It does not affect this patch, but I created #1710 to report that issue. Could you take a look the issue? Thanks.

**Comment by tungld** [comment]
@AlexandreEichenberger could you approve this if it's good? Thanks!

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks for the great work

**Pull request merged**
